wgpu-core/src/lock/validating.rs

```diff
@@ -0,0 +1,163 @@
+  that is about to release a lock.
+
+Thus, the system as a whose is always able to make progress: it is free of
+deadlocks.
+
+Note that this validation only monitors each thread's behavior in isolation:
+there's only thread-local state, nothing communicated between threads. So we
+don't detect deadlocks, per se, only the potential to cause deadlocks. This
+means that the validation is conservative, but more reproducible, since it's not
+dependent on any particular interleaving of execution.
+
+[`lock::rank`]: crate::lock::rank
+*/
+
+#![allow(dead_code)]
+
+use super::rank::{LockRank, LockRankSet};
+use std::{cell::Cell, panic::Location};
+
+pub struct Mutex<T> {
+    inner: parking_lot::Mutex<T>,
+    rank: LockRank,
+}
+
+pub struct MutexGuard<'a, T> {
+    inner: parking_lot::MutexGuard<'a, T>,
+    saved: LockState,
+}
+
+/// Per-thread state for the deadlock checker.
+#[derive(Debug, Copy, Clone)]
+struct LockState {
+    /// The set of locks we are permitted to acquire now.
+    permitted: LockRankSet,
+
+    /// If any locks are held, the location at which the most recent
+    /// lock was acquired.
+    last_acquired: Option<&'static Location<'static>>,
+
+    /// The number of locks currently held.
+    ///
+    /// This is used to enforce stack-like lock acquisition and release.
+    depth: u32,
+}
+
+impl LockState {
+    const INITIAL: LockState = LockState {
+        permitted: LockRank::INITIAL,
+        last_acquired: None,
+        depth: 0,
+    };
+}
+
+impl<T> Mutex<T> {
+    #[inline]
+    pub fn new(rank: LockRank, value: T) -> Mutex<T> {
+        Mutex {
+            inner: parking_lot::Mutex::new(value),
+            rank,
+        }
+    }
+
+    #[inline]
+    #[track_caller]
+    pub fn lock(&self) -> MutexGuard<T> {
+        let state = LOCK_STATE.get();
+        let caller = Location::caller();
+        assert!(
+            state.permitted.contains(self.rank.bit),
+            "Attempt to acquire nested mutexes in wrong order:\n\
+             last acquired a lock at {}\n\
+             now locking at          {}",
+            state.last_acquired.unwrap(),
+            caller,
+        );
+        LOCK_STATE.set(LockState {
+            last_acquired: Some(caller),
+            permitted: self.rank.followers,
+            depth: state.depth + 1,
+        });
+        MutexGuard {
+            inner: self.inner.lock(),
+            saved: state,
+        }
+    }
+}
+
+impl<'a, T> Drop for MutexGuard<'a, T> {
+    fn drop(&mut self) {
+        let prior = LOCK_STATE.replace(self.saved);
+
+        // Although Rust allows mutex guards to be dropped in any
+        // order, this analysis requires that locks be acquired and
+        // released in stack order: the next lock to be released must be
+        // the most recently acquired lock still held.
+        assert_eq!(prior.depth, self.saved.depth + 1,
+                   "Lock not released in stacking order");
```

Comment:
I agree in general, but checking afterwards fits better with the `Cell` API, because we can do a single access with `replace`. My thought was, once a thread has panicked, its state doesn't matter much.

Reply:
A test would need to check the thread's state after a panic, which doesn't seem that interesting.